# Lab | Day 3 Customer Analysis Round 2

* [import pandas + numpy](#1)
* [Standardize header names.](#second-bullet)
* [Drop unnamed column](#third-bullet)
* [Which columns are numerical?](#fourth-bullet)
* [Which columns are categorical?](#fifth-bullet)
* [Check and deal with NaN values.](#sixth-bullet)
* [Deal with NaN values___Sexy nulls snapshot table](#seven-bullet)


## import pandas + numpy <a class="anchor" id="1"></a>


In [9]:
import pandas as pd
import numpy as np

In [10]:
import time
from datetime import date

In [11]:
!ls

marketing_customer_analysis.csv


In [12]:
cd files_for_lab

[Errno 2] No such file or directory: 'files_for_lab'
/Users/AFROARC/Documents/GitHub/BCNDATA1021/Labs/Week1/lab-customer-analysis-round-2-master/files_for_lab/csv_files


In [13]:
cd csv_files

[Errno 2] No such file or directory: 'csv_files'
/Users/AFROARC/Documents/GitHub/BCNDATA1021/Labs/Week1/lab-customer-analysis-round-2-master/files_for_lab/csv_files


In [14]:
df=pd.read_csv('marketing_customer_analysis.csv')
df.head()

Unnamed: 0 Customer       State  Customer Lifetime Value Response  \
0           0  DK49336     Arizona              4809.216960       No   
1           1  KX64629  California              2228.525238       No   
2           2  LZ68649  Washington             14947.917300       No   
3           3  XL78013      Oregon             22332.439460      Yes   
4           4  QA50777      Oregon              9025.067525       No   

   Coverage Education Effective To Date EmploymentStatus Gender  ...  \
0     Basic   College           2/18/11         Employed      M  ...   
1     Basic   College           1/18/11       Unemployed      F  ...   
2     Basic  Bachelor           2/10/11         Employed      M  ...   
3  Extended   College           1/11/11         Employed      M  ...   
4   Premium  Bachelor           1/17/11    Medical Leave      F  ...   

   Number of Open Complaints Number of Policies     Policy Type        Policy  \
0                        0.0                  9  Corporate Auto  Corporate L3   
1                        0.0                  1   Personal Auto   Personal L3   
2                        0.0                  2   Personal Auto   Personal L3   
3                        0.0                  2  Corporate Auto  Corporate L3   
4                        NaN                  7   Personal Auto   Personal L2   

   Renew Offer Type  Sales Channel  Total Claim Amount  Vehicle Class  \
0            Offer3          Agent          292.800000  Four-Door Car   
1            Offer4    Call Center          744.924331  Four-Door Car   
2            Offer3    Call Center          480.000000            SUV   
3            Offer2         Branch          484.013411  Four-Door Car   
4            Offer1         Branch          707.925645  Four-Door Car   

  Vehicle Size Vehicle Type  
0      Medsize          NaN  
1      Medsize          NaN  
2      Medsize            A  
3      Medsize            A  
4      Medsize          NaN  

[5 rows x 26 columns]

In [15]:
# Show data frame shape
df.shape

(10910, 26)

In [16]:
df.columns.to_list()

['Unnamed: 0',
 'Customer',
 'State',
 'Customer Lifetime Value',
 'Response',
 'Coverage',
 'Education',
 'Effective To Date',
 'EmploymentStatus',
 'Gender',
 'Income',
 'Location Code',
 'Marital Status',
 'Monthly Premium Auto',
 'Months Since Last Claim',
 'Months Since Policy Inception',
 'Number of Open Complaints',
 'Number of Policies',
 'Policy Type',
 'Policy',
 'Renew Offer Type',
 'Sales Channel',
 'Total Claim Amount',
 'Vehicle Class',
 'Vehicle Size',
 'Vehicle Type']

## Standardize header names.  <a class="anchor" id="second-bullet"></a>

In [17]:
# change df column names to lowercase.
#replace " " with '_'

def standardize_col(col):
    return col.lower().replace(" ","_")

#check function
standardize_col("income")

'income'

In [18]:
new_cols=[]
for col in df.columns.to_list():
    new_cols.append(standardize_col(col))

In [19]:
new_cols

['unnamed:_0',
 'customer',
 'state',
 'customer_lifetime_value',
 'response',
 'coverage',
 'education',
 'effective_to_date',
 'employmentstatus',
 'gender',
 'income',
 'location_code',
 'marital_status',
 'monthly_premium_auto',
 'months_since_last_claim',
 'months_since_policy_inception',
 'number_of_open_complaints',
 'number_of_policies',
 'policy_type',
 'policy',
 'renew_offer_type',
 'sales_channel',
 'total_claim_amount',
 'vehicle_class',
 'vehicle_size',
 'vehicle_type']

In [20]:
df.columns=new_cols
df

unnamed:_0 customer       state  customer_lifetime_value response  \
0               0  DK49336     Arizona              4809.216960       No   
1               1  KX64629  California              2228.525238       No   
2               2  LZ68649  Washington             14947.917300       No   
3               3  XL78013      Oregon             22332.439460      Yes   
4               4  QA50777      Oregon              9025.067525       No   
...           ...      ...         ...                      ...      ...   
10905       10905  FE99816      Nevada             15563.369440       No   
10906       10906  KX53892      Oregon              5259.444853       No   
10907       10907  TL39050     Arizona             23893.304100       No   
10908       10908  WA60547  California             11971.977650       No   
10909       10909  IV32877         NaN              6857.519928      NaN   

       coverage education effective_to_date employmentstatus gender  ...  \
0         Basic   College           2/18/11         Employed      M  ...   
1         Basic   College           1/18/11       Unemployed      F  ...   
2         Basic  Bachelor           2/10/11         Employed      M  ...   
3      Extended   College           1/11/11         Employed      M  ...   
4       Premium  Bachelor           1/17/11    Medical Leave      F  ...   
...         ...       ...               ...              ...    ...  ...   
10905   Premium  Bachelor           1/19/11       Unemployed      F  ...   
10906     Basic   College            1/6/11         Employed      F  ...   
10907  Extended  Bachelor            2/6/11         Employed      F  ...   
10908   Premium   College           2/13/11         Employed      F  ...   
10909     Basic  Bachelor            1/8/11       Unemployed      M  ...   

       number_of_open_complaints number_of_policies     policy_type  \
0                            0.0                  9  Corporate Auto   
1                            0.0                  1   Personal Auto   
2                            0.0                  2   Personal Auto   
3                            0.0                  2  Corporate Auto   
4                            NaN                  7   Personal Auto   
...                          ...                ...             ...   
10905                        NaN                  7   Personal Auto   
10906                        0.0                  6   Personal Auto   
10907                        0.0                  2  Corporate Auto   
10908                        4.0                  6   Personal Auto   
10909                        0.0                  3   Personal Auto   

             policy  renew_offer_type  sales_channel  total_claim_amount  \
0      Corporate L3            Offer3          Agent          292.800000   
1       Personal L3            Offer4    Call Center          744.924331   
2       Personal L3            Offer3    Call Center          480.000000   
3      Corporate L3            Offer2         Branch          484.013411   
4       Personal L2            Offer1         Branch          707.925645   
...             ...               ...            ...                 ...   
10905   Personal L1            Offer3            Web         1214.400000   
10906   Personal L3            Offer2         Branch          273.018929   
10907  Corporate L3            Offer1            Web          381.306996   
10908   Personal L1            Offer1         Branch          618.288849   
10909   Personal L1            Offer4            Web         1021.719397   

       vehicle_class vehicle_size vehicle_type  
0      Four-Door Car      Medsize          NaN  
1      Four-Door Car      Medsize          NaN  
2                SUV      Medsize            A  
3      Four-Door Car      Medsize            A  
4      Four-Door Car      Medsize          NaN  
...              ...          ...          ...  
10905     Luxury Car      Medsize            A  
10906  Four-Door Car   

## Drop unnamed column  <a class="anchor" id="third-bullet"></a>

In [21]:
dfnew=df.drop(['unnamed:_0'],axis=1)
dfnew

customer       state  customer_lifetime_value response  coverage  \
0      DK49336     Arizona              4809.216960       No     Basic   
1      KX64629  California              2228.525238       No     Basic   
2      LZ68649  Washington             14947.917300       No     Basic   
3      XL78013      Oregon             22332.439460      Yes  Extended   
4      QA50777      Oregon              9025.067525       No   Premium   
...        ...         ...                      ...      ...       ...   
10905  FE99816      Nevada             15563.369440       No   Premium   
10906  KX53892      Oregon              5259.444853       No     Basic   
10907  TL39050     Arizona             23893.304100       No  Extended   
10908  WA60547  California             11971.977650       No   Premium   
10909  IV32877         NaN              6857.519928      NaN     Basic   

      education effective_to_date employmentstatus gender  income  ...  \
0       College           2/18/11         Employed      M   48029  ...   
1       College           1/18/11       Unemployed      F       0  ...   
2      Bachelor           2/10/11         Employed      M   22139  ...   
3       College           1/11/11         Employed      M   49078  ...   
4      Bachelor           1/17/11    Medical Leave      F   23675  ...   
...         ...               ...              ...    ...     ...  ...   
10905  Bachelor           1/19/11       Unemployed      F       0  ...   
10906   College            1/6/11         Employed      F   61146  ...   
10907  Bachelor            2/6/11         Employed      F   39837  ...   
10908   College           2/13/11         Employed      F   64195  ...   
10909  Bachelor            1/8/11       Unemployed      M       0  ...   

      number_of_open_complaints number_of_policies     policy_type  \
0                           0.0                  9  Corporate Auto   
1                           0.0                  1   Personal Auto   
2                           0.0                  2   Personal Auto   
3                           0.0                  2  Corporate Auto   
4                           NaN                  7   Personal Auto   
...                         ...                ...             ...   
10905                       NaN                  7   Personal Auto   
10906                       0.0                  6   Personal Auto   
10907                       0.0                  2  Corporate Auto   
10908                       4.0                  6   Personal Auto   
10909                       0.0                  3   Personal Auto   

             policy  renew_offer_type  sales_channel  total_claim_amount  \
0      Corporate L3            Offer3          Agent          292.800000   
1       Personal L3            Offer4    Call Center          744.924331   
2       Personal L3            Offer3    Call Center          480.000000   
3      Corporate L3            Offer2         Branch          484.013411   
4       Personal L2            Offer1         Branch          707.925645   
...             ...               ...            ...                 ...   
10905   Personal L1            Offer3            Web         1214.400000   
10906   Personal L3            Offer2         Branch          273.018929   
10907  Corporate L3            Offer1            Web          381.306996   
10908   Personal L1            Offer1         Branch          618.288849   
10909   Personal L1            Offer4            Web         1021.719397   

       vehicle_class vehicle_size vehicle_type  
0      Four-Door Car      Medsize          NaN  
1      Four-Door Car      Medsize          NaN  
2                SUV      Medsize            A  
3      Four-Door Car      Medsize            A  
4      Four-Door Car      Medsize          NaN  
...              ...          ...          ...  
10905     Luxury Car      Medsize            A  
10906  Four-Door Car      Medsize            A  
10907     Luxury SUV      Medsize

## Which columns are numerical? <a class="anchor" id="fourth-bullet"></a>

In [22]:
numbercols = dfnew.select_dtypes([np.number])
numbercols.columns.to_list()

['customer_lifetime_value',
 'income',
 'monthly_premium_auto',
 'months_since_last_claim',
 'months_since_policy_inception',
 'number_of_open_complaints',
 'number_of_policies',
 'total_claim_amount']

## Which columns are categorical? <a class="anchor" id="fifth-bullet"></a>

In [23]:
catCols = [col for col in dfnew.columns if df[col].dtype=="O"]
catCols

['customer',
 'state',
 'response',
 'coverage',
 'education',
 'effective_to_date',
 'employmentstatus',
 'gender',
 'location_code',
 'marital_status',
 'policy_type',
 'policy',
 'renew_offer_type',
 'sales_channel',
 'vehicle_class',
 'vehicle_size',
 'vehicle_type']

 ## Check and deal with NaN values. <a class="anchor" id="sixth-bullet"></a>

In [24]:
dfnew.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10910 entries, 0 to 10909
Data columns (total 25 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   customer                       10910 non-null  object 
 1   state                          10279 non-null  object 
 2   customer_lifetime_value        10910 non-null  float64
 3   response                       10279 non-null  object 
 4   coverage                       10910 non-null  object 
 5   education                      10910 non-null  object 
 6   effective_to_date              10910 non-null  object 
 7   employmentstatus               10910 non-null  object 
 8   gender                         10910 non-null  object 
 9   income                         10910 non-null  int64  
 10  location_code                  10910 non-null  object 
 11  marital_status                 10910 non-null  object 
 12  monthly_premium_auto           10910 non-null 

# Deal with NaN values_____Sexy nulls snapshot table   <a class="anchor" id="seven-bullet"></a>

In [25]:
#use this if we think we have any complete row nulls 
dfnew.isna().all(axis=1)

#nulls snapshot table 
nullsdf=pd.DataFrame(round(dfnew.isna().sum()/len(dfnew),4)*100)
nullsdf=nullsdf.rename(columns={'index':'header',0:'proportionnulls'})
nullsdf

proportionnulls
customer                                  0.00
state                                     5.78
customer_lifetime_value                   0.00
response                                  5.78
coverage                                  0.00
education                                 0.00
effective_to_date                         0.00
employmentstatus                          0.00
gender                                    0.00
income                                    0.00
location_code                             0.00
marital_status                            0.00
monthly_premium_auto                      0.00
months_since_last_claim                   5.80
months_since_policy_inception             0.00
number_of_open_complaints                 5.80
number_of_policies                        0.00
policy_type                               0.00
policy                                    0.00
renew_offer_type                          0.00
sales_channel                             0.00
total_claim_amount                        0.00
vehicle_class                             5.70
vehicle_size                              5.70
vehicle_type                             50.25

## Drop Null Values !!Terrible Idea!!        50.25% in 'Vehicle_type'

In [26]:
# dfnew_nonulls=dfnew.dropna()
# dfnew_nonulls.shape

## replace numeric columns - fill in - fill with mean

In [27]:
# numeric columns - fill in - fill with mean
number_of_open_complaints_mean=dfnew["number_of_open_complaints"].mean()

# Replace nulls in 'number_of_open_complaints_mean' with mean !
dfnew['number_of_open_complaints'] = dfnew['number_of_open_complaints'].fillna(number_of_open_complaints_mean)

In [28]:
# numeric columns - fill in - fill with mean
months_since_last_claim_mean=dfnew["months_since_last_claim"].mean()

# Replace nulls in 'months_since_last_claim_mean' with mean !
dfnew['months_since_last_claim'] = dfnew['months_since_last_claim'].fillna(months_since_last_claim_mean)

In [29]:
nullsdf=pd.DataFrame(round(dfnew.isna().sum()/len(dfnew),4)*100)
nullsdf=nullsdf.rename(columns={'index':'header',0:'proportionnulls'})
nullsdf

proportionnulls
customer                                  0.00
state                                     5.78
customer_lifetime_value                   0.00
response                                  5.78
coverage                                  0.00
education                                 0.00
effective_to_date                         0.00
employmentstatus                          0.00
gender                                    0.00
income                                    0.00
location_code                             0.00
marital_status                            0.00
monthly_premium_auto                      0.00
months_since_last_claim                   0.00
months_since_policy_inception             0.00
number_of_open_complaints                 0.00
number_of_policies                        0.00
policy_type                               0.00
policy                                    0.00
renew_offer_type                          0.00
sales_channel                             0.00
total_claim_amount                        0.00
vehicle_class                             5.70
vehicle_size                              5.70
vehicle_type                             50.25

## Extract the months from the dataset and store in a separate column.

In [30]:
dfnew['month']=pd.DatetimeIndex(df['effective_to_date']).month
dfnew.columns

Index(['customer', 'state', 'customer_lifetime_value', 'response', 'coverage',
       'education', 'effective_to_date', 'employmentstatus', 'gender',
       'income', 'location_code', 'marital_status', 'monthly_premium_auto',
       'months_since_last_claim', 'months_since_policy_inception',
       'number_of_open_complaints', 'number_of_policies', 'policy_type',
       'policy', 'renew_offer_type', 'sales_channel', 'total_claim_amount',
       'vehicle_class', 'vehicle_size', 'vehicle_type', 'month'],
      dtype='object')

In [31]:
dfnew.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10910 entries, 0 to 10909
Data columns (total 26 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   customer                       10910 non-null  object 
 1   state                          10279 non-null  object 
 2   customer_lifetime_value        10910 non-null  float64
 3   response                       10279 non-null  object 
 4   coverage                       10910 non-null  object 
 5   education                      10910 non-null  object 
 6   effective_to_date              10910 non-null  object 
 7   employmentstatus               10910 non-null  object 
 8   gender                         10910 non-null  object 
 9   income                         10910 non-null  int64  
 10  location_code                  10910 non-null  object 
 11  marital_status                 10910 non-null  object 
 12  monthly_premium_auto           10910 non-null 

In [32]:
dfnew['effective_to_date']=pd.to_datetime(df['effective_to_date'],errors='coerce')

In [33]:
dfnew['month']=dfnew['effective_to_date'].dt.strftime('%b')

In [34]:
dfnew

customer       state  customer_lifetime_value response  coverage  \
0      DK49336     Arizona              4809.216960       No     Basic   
1      KX64629  California              2228.525238       No     Basic   
2      LZ68649  Washington             14947.917300       No     Basic   
3      XL78013      Oregon             22332.439460      Yes  Extended   
4      QA50777      Oregon              9025.067525       No   Premium   
...        ...         ...                      ...      ...       ...   
10905  FE99816      Nevada             15563.369440       No   Premium   
10906  KX53892      Oregon              5259.444853       No     Basic   
10907  TL39050     Arizona             23893.304100       No  Extended   
10908  WA60547  California             11971.977650       No   Premium   
10909  IV32877         NaN              6857.519928      NaN     Basic   

      education effective_to_date employmentstatus gender  income  ...  \
0       College        2011-02-18         Employed      M   48029  ...   
1       College        2011-01-18       Unemployed      F       0  ...   
2      Bachelor        2011-02-10         Employed      M   22139  ...   
3       College        2011-01-11         Employed      M   49078  ...   
4      Bachelor        2011-01-17    Medical Leave      F   23675  ...   
...         ...               ...              ...    ...     ...  ...   
10905  Bachelor        2011-01-19       Unemployed      F       0  ...   
10906   College        2011-01-06         Employed      F   61146  ...   
10907  Bachelor        2011-02-06         Employed      F   39837  ...   
10908   College        2011-02-13         Employed      F   64195  ...   
10909  Bachelor        2011-01-08       Unemployed      M       0  ...   

      number_of_policies     policy_type        policy  renew_offer_type  \
0                      9  Corporate Auto  Corporate L3            Offer3   
1                      1   Personal Auto   Personal L3            Offer4   
2                      2   Personal Auto   Personal L3            Offer3   
3                      2  Corporate Auto  Corporate L3            Offer2   
4                      7   Personal Auto   Personal L2            Offer1   
...                  ...             ...           ...               ...   
10905                  7   Personal Auto   Personal L1            Offer3   
10906                  6   Personal Auto   Personal L3            Offer2   
10907                  2  Corporate Auto  Corporate L3            Offer1   
10908                  6   Personal Auto   Personal L1            Offer1   
10909                  3   Personal Auto   Personal L1            Offer4   

       sales_channel  total_claim_amount  vehicle_class vehicle_size  \
0              Agent          292.800000  Four-Door Car      Medsize   
1        Call Center          744.924331  Four-Door Car      Medsize   
2        Call Center          480.000000            SUV      Medsize   
3             Branch          484.013411  Four-Door Car      Medsize   
4             Branch          707.925645  Four-Door Car      Medsize   
...              ...                 ...            ...          ...   
10905            Web         1214.400000     Luxury Car      Medsize   
10906         Branch          273.018929  Four-Door Car      Medsize   
10907            Web          381.306996     Luxury SUV      Medsize   
10908         Branch          618.288849            SUV      Medsize   
10909            Web         1021.719397            SUV      Medsize   

      vehicle_type month  
0              NaN   Feb  
1              NaN   Jan  
2                A   Feb  
3                A   Jan  
4              NaN   Jan  
...            ...   ...  
10905            A   Jan  
10906            A   Jan  
10907          NaN   Feb  
10908            A   Feb  
10909          NaN   Jan  

[10910 rows x 26 columns]

# Then filter the data to show only the information for the first quarter , 
 - ie. January, February and March. 
 - **Hint:** If data from March does not exist, consider only January and February.

In [35]:
dfnew.head()

customer       state  customer_lifetime_value response  coverage education  \
0  DK49336     Arizona              4809.216960       No     Basic   College   
1  KX64629  California              2228.525238       No     Basic   College   
2  LZ68649  Washington             14947.917300       No     Basic  Bachelor   
3  XL78013      Oregon             22332.439460      Yes  Extended   College   
4  QA50777      Oregon              9025.067525       No   Premium  Bachelor   

  effective_to_date employmentstatus gender  income  ... number_of_policies  \
0        2011-02-18         Employed      M   48029  ...                  9   
1        2011-01-18       Unemployed      F       0  ...                  1   
2        2011-02-10         Employed      M   22139  ...                  2   
3        2011-01-11         Employed      M   49078  ...                  2   
4        2011-01-17    Medical Leave      F   23675  ...                  7   

      policy_type        policy  renew_offer_type  sales_channel  \
0  Corporate Auto  Corporate L3            Offer3          Agent   
1   Personal Auto   Personal L3            Offer4    Call Center   
2   Personal Auto   Personal L3            Offer3    Call Center   
3  Corporate Auto  Corporate L3            Offer2         Branch   
4   Personal Auto   Personal L2            Offer1         Branch   

   total_claim_amount  vehicle_class vehicle_size vehicle_type month  
0          292.800000  Four-Door Car      Medsize          NaN   Feb  
1          744.924331  Four-Door Car      Medsize          NaN   Jan  
2          480.000000            SUV      Medsize            A   Feb  
3          484.013411  Four-Door Car      Medsize            A   Jan  
4          707.925645  Four-Door Car      Medsize          NaN   Jan  

[5 rows x 26 columns]

## Filter only Jan–Mar

In [36]:
filter1 = dfnew.query('month=="Jan" or month=="Feb" or month=="Mar"')
filter1

customer       state  customer_lifetime_value response  coverage  \
0      DK49336     Arizona              4809.216960       No     Basic   
1      KX64629  California              2228.525238       No     Basic   
2      LZ68649  Washington             14947.917300       No     Basic   
3      XL78013      Oregon             22332.439460      Yes  Extended   
4      QA50777      Oregon              9025.067525       No   Premium   
...        ...         ...                      ...      ...       ...   
10905  FE99816      Nevada             15563.369440       No   Premium   
10906  KX53892      Oregon              5259.444853       No     Basic   
10907  TL39050     Arizona             23893.304100       No  Extended   
10908  WA60547  California             11971.977650       No   Premium   
10909  IV32877         NaN              6857.519928      NaN     Basic   

      education effective_to_date employmentstatus gender  income  ...  \
0       College        2011-02-18         Employed      M   48029  ...   
1       College        2011-01-18       Unemployed      F       0  ...   
2      Bachelor        2011-02-10         Employed      M   22139  ...   
3       College        2011-01-11         Employed      M   49078  ...   
4      Bachelor        2011-01-17    Medical Leave      F   23675  ...   
...         ...               ...              ...    ...     ...  ...   
10905  Bachelor        2011-01-19       Unemployed      F       0  ...   
10906   College        2011-01-06         Employed      F   61146  ...   
10907  Bachelor        2011-02-06         Employed      F   39837  ...   
10908   College        2011-02-13         Employed      F   64195  ...   
10909  Bachelor        2011-01-08       Unemployed      M       0  ...   

      number_of_policies     policy_type        policy  renew_offer_type  \
0                      9  Corporate Auto  Corporate L3            Offer3   
1                      1   Personal Auto   Personal L3            Offer4   
2                      2   Personal Auto   Personal L3            Offer3   
3                      2  Corporate Auto  Corporate L3            Offer2   
4                      7   Personal Auto   Personal L2            Offer1   
...                  ...             ...           ...               ...   
10905                  7   Personal Auto   Personal L1            Offer3   
10906                  6   Personal Auto   Personal L3            Offer2   
10907                  2  Corporate Auto  Corporate L3            Offer1   
10908                  6   Personal Auto   Personal L1            Offer1   
10909                  3   Personal Auto   Personal L1            Offer4   

       sales_channel  total_claim_amount  vehicle_class vehicle_size  \
0              Agent          292.800000  Four-Door Car      Medsize   
1        Call Center          744.924331  Four-Door Car      Medsize   
2        Call Center          480.000000            SUV      Medsize   
3             Branch          484.013411  Four-Door Car      Medsize   
4             Branch          707.925645  Four-Door Car      Medsize   
...              ...                 ...            ...          ...   
10905            Web         1214.400000     Luxury Car      Medsize   
10906         Branch          273.018929  Four-Door Car      Medsize   
10907            Web          381.306996     Luxury SUV      Medsize   
10908         Branch          618.288849            SUV      Medsize   
10909            Web         1021.719397            SUV      Medsize   

      vehicle_type month  
0              NaN   Feb  
1              NaN   Jan  
2                A   Feb  
3                A   Jan  
4              NaN   Jan  
...            ...   ...  
10905            A   Jan  
10906            A   Jan  
10907          NaN   Feb  
10908            A   Feb  
10909          NaN   Jan  

[10910 rows x 26 columns]

# BONUS: 

# ~$\color{red}{\text{Put all the previously mentioned data transformations into a function.}}$~